code/modules/mob/living/simple_animal/hostile/terror_spiders/hive.dm

```diff
@@ -20,21 +21,41 @@
 
 /mob/living/simple_animal/hostile/poison/terror_spider/proc/CountSpiders()
 	var/numspiders = 0
-	for(var/mob/living/simple_animal/hostile/poison/terror_spider/T in GLOB.ts_spiderlist)
+	for(var/thing in GLOB.ts_spiderlist)
+		var/mob/living/simple_animal/hostile/poison/terror_spider/T = thing
 		if(T.stat != DEAD && !T.spider_placed && spider_awaymission == T.spider_awaymission)
 			numspiders += 1
 	return numspiders
 
-/mob/living/simple_animal/hostile/poison/terror_spider/proc/CountSpidersType(specific_type)
+/mob/living/simple_animal/hostile/poison/terror_spider/proc/CountSpidersDetailed(check_mine = FALSE, list/mytypes)
 	var/numspiders = 0
-	for(var/mob/living/simple_animal/hostile/poison/terror_spider/T in GLOB.ts_spiderlist)
-		if(T.stat != DEAD && !T.spider_placed && spider_awaymission == T.spider_awaymission)
-			if(T.type == specific_type)
-				numspiders += 1
-	for(var/obj/structure/spider/eggcluster/terror_eggcluster/E in GLOB.ts_egg_list)
-		if(E.spiderling_type == specific_type && E.z == z)
-			numspiders += E.spiderling_number
-	for(var/obj/structure/spider/spiderling/terror_spiderling/L in GLOB.ts_spiderling_list)
-		if(!L.stillborn && L.grow_as == specific_type && L.z == z)
-			numspiders += 1
+	var/check_list = islist(mytypes) && mytypes.len > 0
+	for(var/thing in GLOB.ts_spiderlist)
+		var/mob/living/simple_animal/hostile/poison/terror_spider/T = thing
+		if(T.stat == DEAD || T.spider_placed || spider_awaymission != T.spider_awaymission)
+			continue
+		if(check_mine && (!T.spider_myqueen || T.spider_myqueen != src))
+			continue
+		if(check_list && !(T.type in mytypes))
```

Comment:
Performance wise.
You can build a type cache here and use that. Assoc list accessing is done in constant time. Where this takes linear time with the length of the list.
Seeing this again as a bonus though due to how little the proc is used.